File: backend/rag_engine.py
from typing import List, Dict

class EnterpriseRAGEngine:
    def __init__(self):
        self.documents = [
            {
                "title": "Sitecore XM Cloud",
                "category": "CMS",
                "content": "Sitecore XM Cloud is a SaaS-based headless CMS platform for enterprise digital experience delivery."
            },
            {
                "title": "RAG Architecture",
                "category": "AI",
                "content": "Retrieval-Augmented Generation combines retrieval systems with large language models to improve contextual enterprise responses."
            },
            {
                "title": "Headless CMS",
                "category": "Architecture",
                "content": "Headless CMS separates frontend delivery from content management using APIs."
            }
        ]

    def semantic_search(self, query: str) -> List[Dict]:
        query = query.lower()
        results = []

        for doc in self.documents:
            score = 0
            content = f"{doc['title']} {doc['category']} {doc['content']}".lower()

            for word in query.split():
                if word in content:
                    score += 1

            if score > 0:
                results.append({
                    "document": doc,
                    "score": score
                })

        return sorted(results, key=lambda x: x['score'], reverse=True)

    def generate_response(self, query: str) -> Dict:
        matches = self.semantic_search(query)

        if not matches:
            return {
                "answer": "No enterprise knowledge found.",
                "sources": []
            }

        context = " ".join([
            item['document']['content'] for item in matches[:3]
        ])

        return {
            "query": query,
            "answer": f"Enterprise AI Response: {context}",
            "sources": [
                item['document']['title'] for item in matches[:3]
            ]
        }


if __name__ == "__main__":
    engine = EnterpriseRAGEngine()

    response = engine.generate_response(
        "Explain Sitecore XM Cloud and RAG architecture"
    )

    print(response)

File: src/lib/rag-engine.js
const enterpriseKnowledge = [
  {
    id: 1,
    title: "Sitecore XM Cloud",
    category: "CMS",
    content: "Sitecore XM Cloud is a SaaS-based headless CMS used for composable digital experience platforms."
  },
  {
    id: 2,
    title: "Headless CMS",
    category: "Architecture",
    content: "Headless CMS separates content management from frontend delivery using APIs such as GraphQL or REST."
  },
  {
    id: 3,
    title: "RAG Architecture",
    category: "AI",
    content: "Retrieval-Augmented Generation improves AI answers by retrieving relevant enterprise knowledge before generating a response."
  },
  {
    id: 4,
    title: "Multilingual Content",
    category: "Content Operations",
    content: "Enterprise CMS platforms often require multilingual workflows, localization, fallback language handling, and SEO-aware routing."
  }
];

function normalize(text) {
  return text.toLowerCase().replace(/[^a-z0-9 ]/g, "");
}

function searchKnowledge(query) {
  const cleanQuery = normalize(query);

  return enterpriseKnowledge
    .map(item => {
      const combinedText = normalize(`${item.title} ${item.category} ${item.content}`);
      const score = cleanQuery
        .split(" ")
        .filter(word => word.length > 2 && combinedText.includes(word)).length;

      return { ...item, score };
    })
    .filter(item => item.score > 0)
    .sort((a, b) => b.score - a.score);
}

function generateRagResponse(query) {
  const results = searchKnowledge(query);

  if (results.length === 0) {
    return {
      answer: "I could not find matching enterprise knowledge. Try asking about Sitecore, RAG, Headless CMS, or multilingual content.",
      sources: []
    };
  }

  const context = results.slice(0, 3).map(item => item.content).join(" ");

  return {
    answer: `Based on enterprise knowledge: ${context}`,
    sources: results.slice(0, 3).map(item => ({
      title: item.title,
      category: item.category,
      score: item.score
    }))
  };
}

module.exports = {
  searchKnowledge,
  generateRagResponse
};
